Update package data paths to use `pathlib`

tbowers7 opened the pull request:
The major focus of this PR is to shift the package data paths to use `pathlib` rather than strings.  In the process, since the core team has decided to deprecate use of python 3.8, the structure of the `data.Paths` class could be updated to use the dual-decorator feature new in python 3.9.

In the process of propagating these changes into the remainder of the codebase, various older constructs were modernized, and import statements were cleaned up (partially to help remove additional `pytest` warnings).

Details:
* Align package version requirements in `setup.cfg`, `environment.yml`, and `tox.ini`.
* With the shift to `pathlib`, the `import os` could be removed in many instances.
* Updated the `os.path` constructs to `pathlib` constructs when connected with `data.Paths` objects.
* A not insignificant amount of import cleaning.

Unit tests pass... dev-suite run after review.

tbowers7 committed:
Update internal data paths to use `pathlib`

Shift from `os.path.join()` usage to `pathlib` for all internal data file
locations.  This allows for all of the functionality of `pathlib` objects and
allows the discontinuance of `import os` in most cases.

Enforce `python>=3.9` in `setup.cfg` and `environment.yml`, which allows the
simplification of the `data.Paths` class to use the dual decorators of
`@classmethod` and `@property` without having to resort to a metaclass
structure.

Various modernization of code happened upon in the process of making the above
changes, along with cleaning up various `pytest` warnings.

	modified:   environment.yml
	modified:   pypeit/core/convert_DEIMOSsavfiles.py
	modified:   pypeit/core/extract.py
	modified:   pypeit/core/findobj_skymask.py
	modified:   pypeit/core/flat.py
	modified:   pypeit/core/flexure.py
	modified:   pypeit/core/flux_calib.py
	modified:   pypeit/core/procimg.py
	modified:   pypeit/core/skysub.py
	modified:   pypeit/core/telluric.py
	modified:   pypeit/core/wavecal/autoid.py
	modified:   pypeit/core/wavecal/patterns.py
	modified:   pypeit/core/wavecal/templates.py
	modified:   pypeit/core/wavecal/waveio.py
	modified:   pypeit/data/arc_lines/convert_NIST_to_lists.py
	modified:   pypeit/data/utils.py
	modified:   pypeit/scripts/ql_multislit.py
	modified:   pypeit/spectrographs/keck_deimos.py
	modified:   pypeit/spectrographs/keck_hires.py
	modified:   pypeit/spectrographs/keck_lris.py
	modified:   pypeit/spectrographs/vlt_xshooter.py
	modified:   pypeit/tests/test_flexure.py
	modified:   pypeit/tests/test_pkgdata.py
	modified:   pypeit/tests/test_runpypeit.py
	modified:   pypeit/tests/tstutils.py
	modified:   pypeit/utils.py
	modified:   setup.cfg

tbowers7 committed:
Doc update / remove 3.8 from `tox`

	modified:   doc/include/dependencies_table.rst
	modified:   tox.ini

tbowers7 committed:
Shift from `Path.joinpath()` to `/` operator

	modified:   pypeit/core/convert_DEIMOSsavfiles.py
	modified:   pypeit/core/flexure.py
	modified:   pypeit/core/flux_calib.py
	modified:   pypeit/core/telluric.py
	modified:   pypeit/core/wavecal/templates.py
	modified:   pypeit/data/utils.py
	modified:   pypeit/scripts/ql_multislit.py
	modified:   pypeit/spectrographs/keck_deimos.py
	modified:   pypeit/spectrographs/keck_lris.py
	modified:   pypeit/spectrographs/vlt_xshooter.py
	modified:   pypeit/tests/test_flexure.py
	modified:   pypeit/tests/test_pkgdata.py
	modified:   pypeit/tests/tstutils.py

kbwestfall commented on environment.yml:
```diff
@@ -3,13 +3,13 @@ channels:
   - numpy>=1.21
   - scipy>=1.7
-  - scikit-learn>=0.24
+  - scikit-learn>=1.0
```
Nice.

tepickering commented on environment.yml:
```diff
@@ -3,13 +3,13 @@ channels:
   - defaults
 dependencies:
-  - python>=3.8,<3.10
+  - python>=3.9,<3.11
```
i've been using `3.11` in my `pypeit` envs for a few weeks now. no issues so i think we can consider it supported. i'd change this to `<3.12`.

tepickering commented on pypeit/core/extract.py:
```diff
@@ -871,7 +868,7 @@ def fit_profile(image, ivar, waveimg, thismask, spat_img, trace_in, wave,
             spline_flux1[indbad2] = np.median(spline_flux1[np.invert(badpix)])
         # take a 5-pixel median to filter out some hot pixels
-        spline_flux1 = scipy.ndimage.filters.median_filter(spline_flux1,size=5,mode ='reflect')
+        spline_flux1 = scipy.ndimage.median_filter(spline_flux1,size=5,mode ='reflect')
```
space before the last `=` will trip strict codestyle checks.

tepickering commented on setup.cfg:
```diff
@@ -28,21 +28,21 @@ classifiers =
 use_2to3=False
 packages = find:
-python_requires = >=3.8,<3.11
+python_requires = >=3.9,<3.11
```
again, i would tweak `3.11` to `3.12` since `3.11.x` is known to work.

tepickering reviewed (approved):
awesome stuff!🤘 cleans up the code very nicely in a number of places.

tbowers7 committed:
Attempt at fix for Windows/py3.9 CI failure

tbowers7 committed:
Address PR comments; another try at Win/py39 fix

tbowers7 committed:
Add TODO for the Win/py39 fix

kbwestfall commented on pypeit/core/extract.py:
```diff
@@ -4,29 +4,22 @@
-from matplotlib import pyplot as plt
+import scipy.interpolate
+import scipy.ndimage
+import scipy.special
```
Nice clean-up!

kbwestfall commented on pypeit/core/findobj_skymask.py:
```diff
@@ -6,10 +6,11 @@
+import matplotlib.pyplot as plt
 
-from astropy import stats
+import astropy.stats
```
Although the over-use of `as` is a pet-peeve of mine, I'm generally not worried about keeping the full module path in imports.  There are times when it's useful or necessary, but I wouldn't suggest starting to propagate these kinds of changes through the code.  But I would leave what you've done already as is.  Just my opinion.

tbowers7 replied:
Thanks.  I've found in my own work that including the full module path in the code is helpful so as to avoid scrolling back up to remember which package something came from.  I will leave these as-is, but will not be overly zealous in the future.

kbwestfall commented on pypeit/core/flexure.py:
```diff
@@ -1525,9 +1527,9 @@ def qa_plots(self, plot_dir:str, root:str):
-            os.mkdir(qa_dir)
+        qa_dir = pathlib.Path(plot_dir) / 'QA'
+        if not qa_dir.is_dir():
+            qa_dir.mkdir()
```
You might want to include `parents=True` here.  I often do that with `mkdir`, unless I know the parent directory should already exists.

kbwestfall commented on pypeit/data/arc_lines/convert_NIST_to_lists.py:
```diff
@@ -76,13 +69,13 @@ def load_line_list(line):
 
     """
-    line_file = os.path.join(data.Paths.nist, f'{line}_vacuum.ascii')
+    line_file = data.Paths.nist.jopinpath(f'{line}_vacuum.ascii')
```
`jopinpath` -> `joinpath` ?

tbowers7 replied:
Whoops.

kbwestfall commented on pypeit/scripts/ql_multislit.py:
```diff
@@ -5,12 +5,11 @@
 
-import os
 import copy
+import os
```
Is `os` still needed?

tbowers7 replied:
Not anymore... changed line in `print_offset_report()` that was still using it.

tbowers7 replied:
Actually, yes, `os` is still needed.

kbwestfall commented on tox.ini:
```diff
@@ -65,10 +65,11 @@ deps =
 extras =
     test
-    alldeps: pyside2,pyqt5,shapely
+    alldeps: pyside2,pyqt5,shapely,specutils
```
Nice!

kbwestfall reviewed (approved):
Very nice.  A couple of minor things from me.

tbowers7 committed:
PR comments

tbowers7 commented:
Requesting `dev-suite` run now.

tbowers7 committed:
CHANGES

tbowers7 committed:
Merge branch 'develop' into data_utils_update

profxj commented:
Uh-oh, @tbowers7 

Test Summary
--------------------------------------------------------
--- PYTEST PYPEIT UNIT TESTS PASSED  208 passed, 2 skipped, 39 warnings in 196.62s (0:03:16) ---
--- PYTEST UNIT TESTS FAILED  17 failed, 101 passed, 171 warnings in 433.95s (0:07:13) ---
--- PYTEST VET TESTS FAILED  10 failed, 26 passed, 5 skipped, 1 warning in 767.39s (0:12:47) ---
--- PYPEIT DEVELOPMENT SUITE FAILED 20/151 TESTS  ---
Failed tests:
    keck_deimos/600ZD_M_6500 pypeit
    keck_deimos/1200G_M_7750 pypeit
    keck_deimos/830G_M_9000_dither pypeit
    gemini_gmos/GS_HAM_R400_860 pypeit_sensfunc
    keck_deimos/600ZD_tilted pypeit
    keck_deimos/1200B_M_5200 pypeit
    keck_deimos/830G_M_8500 pypeit
    keck_deimos/900ZD_M_6000 pypeit
    keck_deimos/830G_L_8100 pypeit
    keck_deimos/1200B_LVM_5200 pypeit
    keck_deimos/830G_LVM_8400 pypeit
    keck_deimos/900ZD_LVM_5500 pypeit
    keck_lris_red_mark4/long_400_8500_d560 pypeit
    keck_deimos/1200G_M_5500 pypeit
    keck_mosfire/Y_long pypeit_sensfunc
    keck_deimos/830G_M_8100_26 pypeit
    vlt_xshooter/NIR pypeit
    gemini_gmos/GS_HAM_R400_700 pypeit_sensfunc
    shane_kast_blue/600_4310_d55 pypeit_sensfunc
    keck_lris_red_mark4/long_600_10000_d680 pypeit
Skipped tests:
    keck_deimos/600ZD_M_6500 pypeit_flux
    keck_deimos/600ZD_M_6500 pypeit_ql
    keck_deimos/600ZD_M_6500 pypeit_ql
    keck_deimos/830G_M_9000_dither pypeit_coadd_2dspec
    gemini_gmos/GS_HAM_R400_860 pypeit_flux_setup
    gemini_gmos/GS_HAM_R400_860 pypeit_flux
    keck_deimos/830G_M_8500 pypeit_multislit_flexure
    keck_deimos/830G_M_8500 pypeit_collate
    keck_deimos/900ZD_LVM_5500 pypeit_sensfunc
    keck_deimos/900ZD_LVM_5500 pypeit_flux
    keck_mosfire/Y_long pypeit_ql
    gemini_gmos/GS_HAM_R400_700 pypeit_flux
    gemini_gmos/GS_HAM_R400_700 pypeit_coadd_1dspec
    gemini_gmos/GS_HAM_R400_700 pypeit_tellfit
    shane_kast_blue/600_4310_d55 pypeit_flux_setup
    shane_kast_blue/600_4310_d55 pypeit_flux
    shane_kast_blue/600_4310_d55 pypeit_coadd_1dspec
    shane_kast_blue/600_4310_d55 pypeit_ql
    shane_kast_blue/600_4310_d55 pypeit_ql
    shane_kast_blue/600_4310_d55 pypeit_ql
    shane_kast_blue/600_4310_d55 pypeit_ql
    shane_kast_blue/600_4310_d55 pypeit_ql
    shane_kast_blue/600_4310_d55 pypeit_ql
    keck_lris_red_mark4/long_600_10000_d680 pypeit_sensfunc
    keck_lris_red_mark4/long_600_10000_d680 pypeit_ql
Testing Started at 2023-03-27T18:00:11.295297
Testing Completed at 2023-03-28T01:19:31.102039
Total Time: 7:19:19.806742

profxj reviewed (changes requested):
putting a stop sign on until we resolve the tests

tbowers7 committed:
Catch missing pathlib conversions

After a horrific DevSuite run, corrected several instances where the full
conversion to pathlib was not done.

	modified:   pypeit/datamodel.py
	modified:   pypeit/spectrographs/keck_deimos.py
	modified:   pypeit/spectrographs/vlt_xshooter.py

tbowers7 committed:
Catch issue with

profxj commented:
Test Summary
--------------------------------------------------------
--- PYTEST PYPEIT UNIT TESTS PASSED  208 passed, 2 skipped, 39 warnings in 209.83s (0:03:29) ---
--- PYTEST UNIT TESTS FAILED  1 failed, 117 passed, 182 warnings in 922.64s (0:15:22) ---
--- PYTEST VET TESTS PASSED  36 passed, 5 skipped, 5 warnings in 794.80s (0:13:14) ---
--- PYPEIT DEVELOPMENT SUITE FAILED 2/179 TESTS  ---
Failed tests:
    keck_mosfire/Y_long pypeit_ql
    keck_lris_red_mark4/long_600_10000_d680 pypeit_ql
Skipped tests:
Testing Started at 2023-03-31T12:42:52.632008
Testing Completed at 2023-03-31T21:20:36.578660
Total Time: 8:37:43.946652

profxj commented:
@tbowers7 -- I'm willing to let those test failures to as they are about to disappear
in my next QL refactor.  So I am now approving.

Please resolve `CHANGES` and merge.

profxj reviewed (approved):
please proceed

tbowers7 committed:
Merge branch 'develop' into data_utils_update

tbowers7 committed:
Finish allowing 3.11

tbowers7 committed:
Doc update
